framework/frequency_manager.hpp

```diff
@@ -0,0 +1,220 @@
+        }
+
+        while (fscanf(file, "%s", read_file)) {
+            if (strcmp(read_file, "userspace") == 0) {
+                fclose(file);
+                return;
+            }
+        }
+
+        fprintf(stderr, "%s: Cannot find \"userspace\" scaling governor from the file: %s\n", program_invocation_name, scaling_governor_path);
+        exit(EXIT_FAILURE);
+#endif
+    }
+
+public:
+    FrequencyManager() {}
+
+    void initial_setup()
+    {
+#ifdef __linux__
+        /* check if "userspace" frequency governor is available. Not all distributions have it*/
+        check_if_userspace_present();    
+
+        /* record supported max and min frequencies */
+        std::string cpuinfo_max_freq_path{"/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq"};
+        max_frequency_supported = get_frequency_from_file(cpuinfo_max_freq_path);
+
+        std::string cpuinfo_min_freq_path{"/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_min_freq"};
+        min_frequency_supported = get_frequency_from_file(cpuinfo_min_freq_path);
+
+        // populate different frequencies for each test to run
+        populate_frequency_levels();
+
+        // save states
+        for (int cpu = 0; cpu < num_cpus(); cpu++) {
+            //save scaling governor for every cpu
+            std::string scaling_governor_path = BASE_FREQ_PATH;
+            scaling_governor_path += std::to_string(cpu_info[cpu].cpu_number);
+            scaling_governor_path += SCALING_GOVERNOR;
+            per_cpu_initial_scaling_governor.push_back(read_file(scaling_governor_path));
+
+            //save frequency for every cpu
+            std::string initial_scaling_setspeed_frequency_path = BASE_FREQ_PATH;
+            initial_scaling_setspeed_frequency_path += std::to_string(cpu_info[cpu].cpu_number);
+            initial_scaling_setspeed_frequency_path += SCALING_SETSPEED;
+            per_cpu_initial_scaling_setspeed.push_back(read_file(initial_scaling_setspeed_frequency_path));
+
+            //change scaling_governor to userspace to have different frequencies
+            if (write_file(scaling_governor_path, "userspace") == 1) {
+                fprintf(stderr, "%s: cannot write userspace to file \"%s\". User should be root :%m\n", program_invocation_name, scaling_governor_path.c_str());
+                exit(EXIT_NOPERMISSION);
+            }
+        }
+#endif 
+    }
+
+    void change_frequency()
+    {
+#ifdef __linux__
+        current_set_frequency = frequency_levels[frequency_level_idx++ % total_frequency_levels];
+        
+        for (int cpu = 0; cpu < num_cpus(); cpu++) {
+            std::string scaling_setspeed = BASE_FREQ_PATH;
+            scaling_setspeed += std::to_string(cpu_info[cpu].cpu_number);
+            scaling_setspeed += SCALING_SETSPEED;
+            if (write_file(scaling_setspeed, std::to_string(current_set_frequency)) == 1) {
+                fprintf(stderr, "%s: cannot write current_frequency (%d) to the file \"%s\" :%m\n", 
```

Comment:
Nitpick: move the ":" to before the space: `\"%s\": %m`. Repeats.

It's far easier to select text in a terminal application if the ":" is not attached to it.